### DISCUSSION TITLE
Restore the history function of old nicks with a cache system

### DISCUSSION THREAD

**Discussion by Chevels** [Zrips/CMI#7221] [Feature Requests]
Hello !
As Mojang has stopped its API to find the old nicknames of the players, is it possible to implement a caching of the nicknames of the players who have already joined the server so that the function is still functional ?

The CheckNameHistory plugin (https://www.spigotmc.org/resources/checknamehistory.3768/) offers this function but it is not usable for users via the config file; on the other hand, the developer proposes that a third-party plugin uses its cache to display old nicknames (see the "For developer" section on overview page).

Knowing the history of nicknames is very useful on community servers, especially for the staff to know who is who. Restoring it that way would be great !

Have a nice week-end ! :wink:

**Comment by mrfloris** [comment]
One problem is that it is not reliable. There's no way to verify the data. Especially as time goes by.

**Comment by Chevels** [comment]
I don't understand what the problem is with reliability. I'm talking about what EssentialX has been doing for a long time :

> The player is also known by the nicknames: OldNick1, OdlNick2, OldNick3...

The plugin only caches all old nicknames a player has had on the server, so I don't see the problem you are talking about in terms of reliability. We may have misunderstood ^^

...and I find it a shame that cmi doesn't offer it yet because it's a very useful function to not lose track of a player and allow better management of the community.

**Comment by mrfloris** [comment]
There's no official api to check any name against. If some error is on an unofficial site that uses old data, you'd have to "assume" it is correct. It's by definition unreliable, because they only have old data, and can't compare new data against it, and only have new data to track, but don't know who just changed over what uuid over time, as multiple people can have the same name. 

CMI had whowas just as much as essentials had it, essentials might internally track it and show old data, cmi is a different plugin. It can't confirm with mojang services anymore, and so the command whowas has been removed. 

I wouldn't want a production server to mass ban people because whowas info from a third party site can't be checked. not a risk I'm willing to take. and yo never know if those sites who ALSO have to be compliant with directives etc are going to purge data now, or charge for it, etc.

Every server for years have been using UUID, those are how bans are set and checked, as well. Any reasonable plugin should be able to manage a server by just uuid. And should have done for years. There's not much that will change besides the convenience of knowing their name history.

**Comment by Momshroom** [comment]
I think the original request was that CMI report only what the player had used as names while on that server, not get info from a 3rd party site, but merely cache what was seen by the server itself. I would be interested in that as well.  Plugins work by uuid, but people don’t and it is nice for staff to know who they are talking to.

**Comment by Chevels** [comment]
Yes that's exactly it Momshroom^^

I don't know why MrFloris ( :p ) thinks I'm asking for control on a third party site or other stuff that would require an API or something complicated, I'm just asking for the names of the players who join the server are stored by the plugin. I ask nothing more, and in these conditions I repeat that I absolutely do not see the problem since there can be no error.

At the time when CMI was starting and Zrips was still accessible, it was me who made the suggestion to Zrips to use the Mojang API to find out the old names of the players. He liked the idea and implemented it in the next update. The API doesn't work anymore, but it's easy to store locally the names of players who connect to the server, other plugins do, but since I've been supporting CMI for years and it's become a plugin that wants to be very comprehensive, it would be a real shame if it didn't end up implementing this through a local cache.

**Reply by mrfloris** [reply]
> Yes that's exactly it Momshroom^^
> 
> I don't know why MrFloris ( :p ) thinks I'm asking for control on a third party site or other stuff that would require an API or something complicated, I'm just asking for the names of the players who join the server are stored by the plugin. I ask nothing more, and in these conditions I repeat that I absolutely do not see the problem since there can be no error.
> 
> At the time when CMI was starting and Zrips was still accessible, it was me who made the suggestion to Zrips to use the Mojang API to find out the old names of the players. He liked the idea and implemented it in the next update. The API doesn't work anymore, but it's easy to store locally the names of players who connect to the server, other plugins do, but since I've been supporting CMI for years and it's become a plugin that wants to be very comprehensive, it would be a real shame if it didn't end up implementing this through a local cache.

<img width="919" alt="Screen-Shot 2022-09-25 at 23 45 28" src="https://user-images.githubusercontent.com/28841349/192166952-a8f5fdbf-109e-422f-bd1e-1610cdf90720.png">

Well to be fair, after the whowas goes away you ask for a cache of this, I explain why the info is not reliable and that mojang removed the API.. the whowas and other info comes through that, to get the real uuid. 

Your question also mentions "this plugin does that", that plugin literally says in the overview tab "Using Mojang's official API!" 

But yes, a plugin could start tracking the name for the uuid, but as explained, you shouldn't depend on it as there is no way to verify who they claim in their uuid who they were in the past. For obvious reasons. 

I am not against the idea of CMI starting to track internally, as the server already knows the username changes 'uuid IGN x is now logging in with Y', the plugins can use this. 

You say "Zrips was still accessible" <- I don't understand why you imply that he isn't? I speak to him every other day or so through DM, he is active on github in his repos, and he has frequent releases. 

I am not here to fight you, or talk this idea down. We're discussing it .. but saying "I don't know why you thought .. " when your sentence literally implies "looking up old names isn't possible, can we get an alternative", yes, then I talk about alternatives .. you can't then go and imply that I just go on some random topic that has nothing to diwth this topic. I'm replying to what you say. If I misinterpret your intentions, fine, so be it.

no error? What about user "something" who joined in 2011, played until 2014, changed name 3x, and now returns tomorrow after the API is unavailable, but the server had the cmi plugin since 2015? .. There's no cache locally to pull from, there's nothing to check against. And whatever third party services like namemc might claim they have on record, those are cached data - there's no way for namemc to verify their history data is correct.  There are numerous examples like this. 

Thankfully it simply doesn't matter. the server knows when someone's name has changed when they login, a plugin can use that - for THAT moment it's possible to store their uuid with previous and current, appending to some list. 

The next question is though: eu and other laws are partially the reason why mojang removes this, eu directives kinda demand services that store certain data to not store it for longer than x, any plugin that finds a solution for caching username history has to comply - or at least the server owner I guess - to those laws, directorive, etc. at least currently cmi doesn't track personal details into infinity.

Regardless of opinions, server-owner management convenience is a huge reason why I objected against Mojang removing username history. I've brought it up in various discussions. So don't worry, in a way I am on your side on this topic; and hopefully CMI has some sort of usable solution at some point. I am merely sharing that there's no data to verify against as that service is gone. And that one shouldn't assume any data they can get through unofficial third party services are correct.

**Reply by Chevels** [reply]
Don't worry, I want to make it clear that the tone I use in my posts remains friendly and that even if it could give the impression, I have never criticized either CMI or you (hence the smiley :p for example).
I understand what you are talking about of course, and I also understand that a member who changes his name outside the server will not be referenced in the local cache, but in practice it is not important.
For example my server is 12 years old, and our map is just as old even if I try to update it manually in terms of generation (yes, I'm completely crazy haha); such an old map has many traces left by players and this history of the nicknames of the same player is very useful to us because it allows us to make the link between different builds of the same player who has changed his name over years. As our server is partly RP, it is often beneficial to know that such player has connected with such nicknames.
And as generally the UUID does not change, the plugin makes the link between the different nicknames (like LWCX does, otherwise it would be unlivable^^).

At worst, I continue to use EssentialsX in addition to CMI just for that, but it's a shame to have 2 plugins that do the same thing (even if CMI is more complete) just for that :/ And in view of what i read on different forums, i don't think i'm the only one who needs to keep track of old player names.
Now you have the cards in hand, we will see what will be your final decision on this subject.

In any case, thank you for your attention to the question, and have a good week! :)

**Reply by mrfloris** [reply]
Thanks for the extra info and sorry for being short in my tone, my migraine today has been testing me. I have zero influence in the decision making. It's all a one man show for zrips. I think with it being a week after the mojang change, that plugins, sites, servers, etc are still trying to figure it all out. Who knows, maybe zrips is already working on something.
